hs-bindgen/src-internal/HsBindgen/Artefact.hs

```diff
@@ -107,32 +110,44 @@ runArtefacts
     env :: ArtefactEnv
     env = ArtefactEnv tracer
 
-    go :: Artefacts as -> ArtefactM (NP I as)
-    go Nil       = pure Nil
-    go (a :* as) = (:*) . I <$> runArtefact a <*> go as
-
-    runArtefact :: Artefact a -> ArtefactM a
+    go :: Artefacts x -> ArtefactM (Either (TraceException msg) (NP I x))
+    go Nil       = pure (Right Nil)
+    go (a :* as) = do
+      artefactResult <- runArtefact a
+      case artefactResult of
+        Left errors -> pure (Left errors)
+        Right artefact -> do
+          tracerState <- liftIO $ readIORef tracerStateRef
```

Comment:
Can we instead export a function from `Tracer.hs` performing this check, and only use this function here? `runArtefacts` does not need to know how to detect an error, only that an error has happend (with a list of error traces I guess).